### ISSUE TITLE
Implement, export, and document forgeBSgenomeDataPkgFromNCBI()

### ISSUE THREAD

**Issue by hpages** [Bioconductor/BSgenomeForge#11]
To do only once PR #10 is merged. Make sure to resync your fork once PR #10 is merged.

**`forgeBSgenomeDataPkgFromNCBI()` will perform the 4 following steps:**

1. Download the FASTA file for the specified assembly from NCBI.

2. Convert the downloaded FASTA file to the 2bit format, with reordering and renaming of the sequences. The 2bit file should be named `single_sequences.2bit`.

3. Use the `createPackage()` function from the **Biobase** package to create the directory structure of the BSgenome data package. The _package template_ that we're going to use for this is the `inst/pkgtemplates/NCBI_BSgenome_datapkg` directory located in the **BSgenomeForge** package.

4. Copy `single_sequences.2bit` to the `inst/extdata` folder of the forged package.

**`forgeBSgenomeDataPkgFromNCBI()` arguments:**

Mandatory arguments:

- `assembly_accession`: Single string. Assembly accession (GenBank or RefSeq).

- `organism`: Single string. Full organism name e.g. `"Homo sapiens"`, `"Mus musculus"`, `"Felis catus"`, `"Acidianus infernus"`, etc...

- `genome`: Single string. The name of the assembly.

- `pkg_maintainer`: Single string. Must be of the form `"Some name <some@email.address>"`

Optional arguments:

- `pkg_author`: Single string. Default value is `NA`. If not supplied, then should be considered to be the same as the `maintainer`.

- `pkg_version`: Single string. The version of the package e.g. `"0.6.1"`. Default value is `"1.0.0"`.

- `license`: Single string. Default value is `"Artistic-2.0"`.

- `destdir`: Single string. Path to the directory where to create the package. Default value is `"."`.

We're going to start with this but we might add more arguments later. The idea is to collect enough information via the arguments of the function in order to be able to infer the `pkgname` (single string) and `symbolValues` (named list) arguments passed to `createPackage()`. For example `pkgname`  should be `BSgenome.<abbreviated organism>.NCBI.<genome>`.

**Comment by hpages** [comment]
Also feel free to make modifications to the package template if you see opportunities to remove some unnecessary or redundant symbols. For example, as mentioned previously in issue #9, should we get rid of symbols `@PROVIDER@`, `@SUGGESTS@`, and `@PKGDESCRIPTION@`? Also what about `@PKGTITLE@`?

For `@PROVIDER@`, just replace with `NCBI` in the template because we know in advance that this symbol will always be replaced with NCBI in this template (the `NCBI_BSgenome_datapkg` template).

For `@SUGGESTS@`, simply remove the `Suggests: @SUGGESTS@` line in the `DESCRIPTION` file of the template.

For `@PKGDESCRIPTION@`, we could certainly edit `DESCRIPTION` and `man/package.Rd` and replace `@PKGDESCRIPTION@` with a full sentence. Note that the sentence would probably still need to contain some symbols e.g. it could be something like this:
```
Full genome sequences for @ORGANISM@ (assembly accession @ACCESSION@, assembly name @GENOME@) as provided by NCBI and stored in Biostrings objects.
```
However, note that this would introduce a new symbol (`@ACCESSION@`), so we're not really getting rid of a symbol by doing this. Also this means that we need to make the same change in 2 different places (`DESCRIPTION` and `man/package.Rd`), and that we need to remember to edit those 2 places again in the future to keep them in sync if we ever decide to slightly modify this line. This doesn't sound as flexible as constructing the `PKGDESCRIPTION` string in R, and injecting it in the package to forge via `createPackage()`.

I could make the same comment about `@PKGTITLE@`.

So I'm not convinced about getting rid of `@PKGDESCRIPTION@` or ``@PKGTITLE@` but feel free to try and see how it goes.

**Comment by kakopo** [comment]
Hi @hpages. I've tried to work on the function, and all is well till it comes across the `symValues` object, to which I get the following error when testing the `forgeBSgenomeDataPkgFromNCBI()` script in an R session:
```Error in copySubstitute(dir(originDir, full.names = TRUE), pkgdir, symbolValues,  : 
  'symbolValues' must have non-empty names.
```
I've tried to debug the code to see what is going wrong but I don't see where I may have gone astray with the list in `symValues`. I'd really appreciate some help going through this part of the code.

Otherwise, yes, after your explanation I agree that changing `@PKGDESCRIPTION@` and `@PKGTITLE@` would be less flexible in future maintenance, so perhaps the current approach is best (although I suspect my approach at these two symbols is the cause of the former error message). 

Returning to the issue of `symValues`, I wonder whether `CIRCSEQS`  in the `zzz.R ` file should be specified as an argument, as some organisms have circular sequences and other do not. Perhaps a default `NA` value? 

I also employed use of some approximations where I wasn't sure how to approach the problem, such as `BSGENOMEOBJNAME = abbr`, because it seems to be an abbreviation of the organism name in most instances. Is this alright or is there another possible alternative?

I am still trying to improve the file but this is where I am currently stuck. I'd appreciate a few pointers. Thanks!

**Comment by hpages** [comment]
Hi @kakopo,

So you have this long and complex and hard to read statement that you use to create the `symValues` list:
```
    symValues <- list(PKGNAME = pkgname, BSGENOMEOBJNAME = abbr,
                      PKGTITLE = "Full genome sequences for", organism, "(NCBI version", genome,")",
                      PKGAUTHOR = pkg_author, VERSION = pkg_version, MAINTAINER = pkg_maintainer,
                      PKGDESCRIPTION = "Full genome sequences for", organism, "as provided by
                      NCBI (", genome, ") and stored in Biostrings objects.", LICENSE = "Artistic-2.0",
                      ORGANISM = organism, GENOME = genome, ORGANISMBIOCVIEW = genome,
                      SEQNAMES = "NA", CIRCSEQS = "NA")
```
The first thing I would suggest is that you format it in a more readable way e.g. by using one line per list element:
```
    symValues <- list(
        PKGNAME = pkgname,
        BSGENOMEOBJNAME = abbr,
        PKGTITLE = "Full genome sequences for", organism, "(NCBI version", genome,")",
        PKGAUTHOR = pkg_author,
        VERSION = pkg_version,
        MAINTAINER = pkg_maintainer,
        PKGDESCRIPTION = "Full genome sequences for", organism, "as provided by NCBI (", genome, ") and stored in Biostrings objects.",
        LICENSE = "Artistic-2.0",
        ORGANISM = organism,
        GENOME = genome,
        ORGANISMBIOCVIEW = genome,
        SEQNAMES = "NA",
        CIRCSEQS = "NA"
    )
```
Isn't that easier to read?

Now there is one obvious problems with this. For example when you set the `PKGTITLE` symbol value with:
```
        ...
        PKGTITLE = "Full genome sequences for", organism, "(NCBI version", genome,")",
        ...
```
does this really do what you think it does?

To see what happens exactly, try to construct a small named list with:
```
organism <- "Felis catus"
genome <- "Felis_catus_9.0"

symValues <- list(
    A = "foo",
    PKGTITLE = "Full genome sequences for", organism, "(NCBI version", genome, ")",
    C = "bar"
)
```
Then take a close look at `symValues`. How many elements does this list have? What are the names on the list? (extract them with `names(symValues)`).

Can you see what happened?

You seem to expect that R will automatically concatenate the 5 strings `"Full genome sequences for"`, `organism`, `"(NCBI version"`, `genome`, and `")"`. But R does not do that. Instead each string is considered to be a separate list element. The following reformatting of the above code reflects better what R does:
```
symValues <- list(
    A = "foo",
    PKGTITLE = "Full genome sequences for",
    organism,
    "(NCBI version",
    genome,
    ")",
    C = "bar"
)
```

To concatenate strings, you must use `paste()` or `paste0()`. But I would suggest that you do this before you construct `symValues` e.g.:
```
organism <- "Felis catus"
genome <- "Felis_catus_9.0"
pkgtitle <- paste0("Full genome sequences for ", organism, " (NCBI version ", genome, ")")
symValues <- list(
    A = "foo",
    PKGTITLE = pkgtitle,
    C = "bar"
```

Let's set `CIRCSEQS` to `"character(0)"` for now (this means "no circular sequence"). Yes we will need to add a `circ_seqs` argument to the `forgeBSgenomeDataPkgFromNCBI()` function, but we'll do this later, as a separate task.

**Comment by kakopo** [comment]
Hi @hpages. Thank you for the response. It helped me understand string concatenation (and when to use the `paste0()` function) much better. 
I've updated the code from in the PR- kindly have a look and let me know if it is alright. 
I also realized I didn't fully understand this part of the task:

> Use the createPackage() function from the Biobase package to create the directory structure of the BSgenome data package. The package template that we're going to use for this is the inst/pkgtemplates/NCBI_BSgenome_datapkg directory located in the BSgenomeForge package.

Is the inst file, in this case, meant to be within the package a user shall download, or is it the package file currently being worked on? I interpreted it as the former, and copied the inst file to the downloaded `BSgenomeForge` package, but would like a bit more clarity.

**Comment by hpages** [comment]
Remember that the purpose of the function is to produce a valid BSgenome data package so you want to make sure that it does that.

For example, if we want to use it to forge a BSgenome data package for the _Acidianus infernus_ assembly used in the `downloadGenomicSequencesFromNCBI()` and `fastaTo2bit()` examples, then we're going to call the function with the following arguments:

- `assembly_accession`: `"GCA_009729545.1"` (or `"GCF_009729545.1"`)

- `organism`: `"Acidianus infernus"`

- `genome`: `"ASM972954v1"`

- `pkg_maintainer`: `"Some name <some@email.address>"`

This should produce a package named `BSgenome.Ainfernus.NCBI.ASM972954v1`.

To check that the package is valid, proceed as follow:

1. Perform some basic visual inspection e.g. look at its directory structure (with the `tree` command) and make sure that looks as expected, that is, that the _top directory_ contains a `DESCRIPTION` and a `NAMESPACE` file as well as an `R/` and a `man/` folder. Don't hesitate to compare to what you've obtained in October/November when you created BSgenome data packages for Cat. The package source tree should be the same.

    This will clear out your doubts about whether you should use `inst/pkgtemplates` or  `inst/pkgtemplates/NCBI_BSgenome_datapkg` as the package template when calling `createPackage()`. One won't produce a valid package source tree because it will add an extra level of directory nesting to it which means that the `DESCRIPTION` file and all the things that we expect to see in the top level folder won't be there (they will be in a subfolder). 

    It's also worth doing some visual inspection of the `DESCRIPTION` file itself. Does it look as expected?

2. Then, try to install the `BSgenome.Ainfernus.NCBI.ASM972954v1` package (with `R CMD build` followed by `R CMD INSTALL` on the source tarball). Load it in a fresh R session and play with it (ad hoc manual testing).

3. Finally run `R CMD check` on the source tarball.

Hope this helps.

**Comment by hpages** [comment]
The man page for `system.file()` is probably one of the worst. First of all it's not really about finding names it's about finding paths. Second, it wouldn't hurt if the Description section said something like _Finds **absolute** paths of files located in **installed** packages_ as this would avoid any potential confusion. Third, they never mention "the `inst` gotcha". Finally the name of the function itself is not really helping. IMO something like `package.file()` would be a better name.

**Some terminology:**

- _package source tree_: This is the directory that gets created on your machine when you clone a package from GitHub or from other remote repository. This is where you do your development work on the package, typically by editing/adding/removing files. This is also what you run `R CMD build` on.

- `R CMD build` produces a _package source tarball_ (a.k.a. _source package_). The _source package_ always has the `.tar.gz` suffix. You can display its content with `tar ztf <path/to/the/source/tarball>`. This will show the same directory structure as in the  _package source tree_ plus some additional folders and files.

- _package installation directory_: This is the directory where the package gets installed when you run `R CMD INSTALL` on the source tarball produced by `R CMD build`. R controls the location and the content of the _package installation directory_ so we should never try to modify its content manually. More generally speaking, the _package installation directory_ should always be treated as read-only. In other words, once a package is installed, the content of its installation directory should never be touched. It will only be touched by R itself when we reinstall the package.
  `system.file(package="toto")` returns the path to the installation folder of package **toto**.

**About "the `inst` gotcha":**

When we install a package, the directory structure located inside the `inst/` folder of the _package source tree_ (or the `source package`) gets copied to the _top-level_ folder of the _package installation directory_.

For example, `inst/foo/bar` will be copied as `foo/bar`. The best way to see this is to try something like `dir(system.file(package="BSgenomeForge"), recursive=TRUE)`. This will list all the files that got installed in the installation folder of package **BSgenomeForge**.

This means that if we want to construct the path to something located under `inst/`, we should _not_ specify `"inst"` in our call to `system.file()`. For example:
```
system.file(package="BSgenomeForge", "inst", "pkgtemplates", "NCBI_BSgenome_datapkg")  # NO!
system.file(package="BSgenomeForge", "pkgtemplates", "NCBI_BSgenome_datapkg")          # YES!
```
Hope this helps.

**Comment by kakopo** [comment]
This definitely helped! I was at a loss but now I have a better understanding of what's involved when a package is installed (and its tree structure). I have updated the code in the PR, and added a man page- please take a look and let me know if they are alright.

**Comment by hpages** [comment]
I merged PR #12 so we can close this.

FYI, after merging, I made various small tweaks to `forgeBSgenomeDataPkgFromNCBI()`, and also bumped the version of the package to 0.0.5. See commit 2d24b7cf89f55858c29d812de252c901795e59d6 for the details. Before you do any further work on your fork, make sure to resync it (and to `git pull --all` in your local clone after the resync) .

Next thing we need to work on is add the `circ_seqs` argument to the function. This is issue #13. See you there!
